File: frontend/src/lib/warehouseOperationPolicy.ts
/**
 * Central warehouse requirement policy — OMS workflow vs WMS/stock ops.
 *
 * Use {@link getOperationPolicy} / {@link requiresWarehouse} everywhere instead of
 * ad-hoc `if (!selectedWarehouse)` checks.
 */

/** Domain context of an operation (future-proof classification). */
export enum OperationContext {
  NONE = "NONE",
  ORDER_WORKFLOW = "ORDER_WORKFLOW",
  INVENTORY = "INVENTORY",
  LOCATION = "LOCATION",
  DOCUMENT = "DOCUMENT",
  RESERVATION = "RESERVATION",
  PRODUCTION = "PRODUCTION",
}

/** Canonical operation ids used by FE + BE policy (keep in sync with warehouse_operation_policy.py). */
export type WarehouseOperation =
  // ORDER_WORKFLOW — no warehouse
  | "order.change_panel_status"
  | "order.change_status"
  | "order.change_customer"
  | "order.change_tags"
  | "order.remove_tags"
  | "order.set_priority"
  | "order.add_note"
  | "order.delete_note"
  | "order.change_payment_status"
  | "order.change_shipping"
  | "order.issue_document_type"
  | "order.edit"
  | "order.export"
  | "order.print_documents"
  | "order.custom_field_value"
  | "order.delete_orders"
  // DOCUMENT (warehouse / fiscal stock docs) — warehouse required
  | "document.create_warehouse"
  | "document.delete_warehouse"
  // RESERVATION
  | "reservation.create"
  | "reservation.release"
  | "reservation.delete"
  // INVENTORY / WMS stock
  | "wms.picking"
  | "wms.picking_bulk"
  | "wms.packing"
  | "wms.receiving"
  | "wms.putaway"
  | "wms.putaway_bulk"
  | "wms.mm"
  | "wms.inventory_count"
  | "wms.relocation"
  | "wms.returns_terminal"
  | "wms.delete_picking_batch"
  // LOCATION
  | "location.create"
  | "location.update"
  | "location.delete"
  | "location.move_stock"
  // PRODUCTION
  | "production.execute"
  | "production.reserve_materials"
  | "admin.warehouse_settings";

export type OperationPolicy = {
  requiresWarehouse: boolean;
  context: OperationContext;
  reason: string;
};

const POLICY: Record<WarehouseOperation, OperationPolicy> = {
  "order.change_panel_status": {
    requiresWarehouse: false,
    context: OperationContext.ORDER_WORKFLOW,
    reason: "Status panelu to workflow OMS, nie operacja na stanach.",
  },
  "order.change_status": {
    requiresWarehouse: false,
    context: OperationContext.ORDER_WORKFLOW,
    reason: "Status zamówienia to workflow OMS.",
  },
  "order.change_customer": {
    requiresWarehouse: false,
    context: OperationContext.ORDER_WORKFLOW,
    reason: "Dane klienta nie zależą od magazynu.",
  },
  "order.change_tags": {
    requiresWarehouse: false,
    context: OperationContext.ORDER_WORKFLOW,
    reason: "Tagi to metadane zamówienia.",
  },
  "order.remove_tags": {
    requiresWarehouse: false,
    context: OperationContext.ORDER_WORKFLOW,
    reason: "Usuwanie tagów to metadane zamówienia.",
  },
  "order.set_priority": {
    requiresWarehouse: false,
    context: OperationContext.ORDER_WORKFLOW,
    reason: "Priorytet to atrybut workflow.",
  },
  "order.add_note": {
    requiresWarehouse: false,
    context: OperationContext.ORDER_WORKFLOW,
    reason: "Notatka nie modyfikuje stanów.",
  },
  "order.delete_note": {
    requiresWarehouse: false,
    context: OperationContext.ORDER_WORKFLOW,
    reason: "Usuwanie notatki nie modyfikuje stanów.",
  },
  "order.change_payment_status": {
    requiresWarehouse: false,
    context: OperationContext.ORDER_WORKFLOW,
    reason: "Status płatności to workflow OMS.",
  },
  "order.change_shipping": {
    requiresWarehouse: false,
    context: OperationContext.ORDER_WORKFLOW,
    reason: "Metoda dostawy to atrybut zamówienia.",
  },
  "order.issue_document_type": {
    requiresWarehouse: false,
    context: OperationContext.ORDER_WORKFLOW,
    reason: "Ustawienie typu dokumentu (flaga), nie wystawienie dokumentu magazynowego.",
  },
  "order.edit": {
    requiresWarehouse: false,
    context: OperationContext.ORDER_WORKFLOW,
    reason: "Edycja pól OMS bez ruchu towaru.",
  },
  "order.export": {
    requiresWarehouse: false,
    context: OperationContext.ORDER_WORKFLOW,
    reason: "Eksport danych jest operacją biurową.",
  },
  "order.print_documents": {
    requiresWarehouse: false,
    context: OperationContext.ORDER_WORKFLOW,
    reason: "Druk etykiet/dokumentów handlowych nie wymaga aktywnego magazynu WMS.",
  },
  "order.custom_field_value": {
    requiresWarehouse: false,
    context: OperationContext.ORDER_WORKFLOW,
    reason: "Pola dodatkowe to metadane zamówienia.",
  },
  "order.delete_orders": {
    requiresWarehouse: false,
    context: OperationContext.ORDER_WORKFLOW,
    reason: "Usuwanie/archiwizacja rekordu zamówienia to operacja administracyjna OMS, nie ruch towaru.",
  },

  "document.create_warehouse": {
    requiresWarehouse: true,
    context: OperationContext.DOCUMENT,
    reason: "Dokument magazynowy (PZ/WZ/MM) jest zawsze w kontekście magazynu.",
  },
  "document.delete_warehouse": {
    requiresWarehouse: true,
    context: OperationContext.DOCUMENT,
    reason: "Usuwanie dokumentu magazynowego wymaga magazynu dokumentu.",
  },

  "reservation.create": {
    requiresWarehouse: true,
    context: OperationContext.RESERVATION,
    reason: "Rezerwacja blokuje stany w magazynie.",
  },
  "reservation.release": {
    requiresWarehouse: true,
    context: OperationContext.RESERVATION,
    reason: "Zwolnienie rezerwacji dotyczy stanów magazynowych.",
  },
  "reservation.delete": {
    requiresWarehouse: true,
    context: OperationContext.RESERVATION,
    reason: "Usuwanie rezerwacji dotyczy stanów magazynowych.",
  },

  "wms.picking": {
    requiresWarehouse: true,
    context: OperationContext.INVENTORY,
    reason: "Zbieranie schodzi ze stanów lokalizacji.",
  },
  "wms.picking_bulk": {
    requiresWarehouse: true,
    context: OperationContext.INVENTORY,
    reason: "Masowe zbieranie to proces WMS na stanach.",
  },
  "wms.packing": {
    requiresWarehouse: true,
    context: OperationContext.INVENTORY,
    reason: "Pakowanie operuje na zebranych stanach / nośnikach.",
  },
  "wms.receiving": {
    requiresWarehouse: true,
    context: OperationContext.INVENTORY,
    reason: "Przyjęcie zwiększa stany magazynowe.",
  },
  "wms.putaway": {
    requiresWarehouse: true,
    context: OperationContext.LOCATION,
    reason: "Rozlokowanie umieszcza towar na lokalizacjach.",
  },
  "wms.putaway_bulk": {
    requiresWarehouse: true,
    context: OperationContext.LOCATION,
    reason: "Masowe rozlokowanie to proces WMS na lokalizacjach.",
  },
  "wms.mm": {
    requiresWarehouse: true,
    context: OperationContext.LOCATION,
    reason: "Przesunięcie MM zmienia lokalizacje i stany.",
  },
  "wms.inventory_count": {
    requiresWarehouse: true,
    context: OperationContext.INVENTORY,
    reason: "Inwentaryzacja koryguje stany.",
  },
  "wms.relocation": {
    requiresWarehouse: true,
    context: OperationContext.LOCATION,
    reason: "Rozlokowanie nośników to operacja lokalizacyjna.",
  },
  "wms.returns_terminal": {
    requiresWarehouse: true,
    context: OperationContext.INVENTORY,
    reason: "Terminal zwrotów przyjmuje towar na stany.",
  },
  "wms.delete_picking_batch": {
    requiresWarehouse: true,
    context: OperationContext.INVENTORY,
    reason: "Usuwanie zbioru zbierania dotyczy artefaktów WMS w magazynie.",
  },

  "location.create": {
    requiresWarehouse: true,
    context: OperationContext.LOCATION,
    reason: "Lokalizacja należy do magazynu.",
  },
  "location.update": {
    requiresWarehouse: true,
    context: OperationContext.LOCATION,
    reason: "Edycja lokalizacji w magazynie.",
  },
  "location.delete": {
    requiresWarehouse: true,
    context: OperationContext.LOCATION,
    reason: "Usuwanie lokalizacji w magazynie.",
  },
  "location.move_stock": {
    requiresWarehouse: true,
    context: OperationContext.LOCATION,
    reason: "Ruch towaru między lokalizacjami.",
  },

  "production.execute": {
    requiresWarehouse: true,
    context: OperationContext.PRODUCTION,
    reason: "Wykonanie produkcji zużywa/przyjmuje stany.",
  },
  "production.reserve_materials": {
    requiresWarehouse: true,
    context: OperationContext.PRODUCTION,
    reason: "Rezerwacja surowców na produkcję.",
  },
  "admin.warehouse_settings": {
    requiresWarehouse: true,
    context: OperationContext.NONE,
    reason: "Konfiguracja jest per magazyn.",
  },
};

/** Full policy for an operation. */
export function getOperationPolicy(operation: WarehouseOperation): OperationPolicy {
  return POLICY[operation];
}

/** Convenience: whether warehouse context is mandatory. */
export function requiresWarehouse(operation: WarehouseOperation): boolean {
  return getOperationPolicy(operation).requiresWarehouse;
}

/** Alias for callers that think in “context” terms. */
export function requiresWarehouseForContext(context: OperationContext): boolean {
  switch (context) {
    case OperationContext.NONE:
    case OperationContext.ORDER_WORKFLOW:
      return false;
    case OperationContext.INVENTORY:
    case OperationContext.LOCATION:
    case OperationContext.DOCUMENT:
    case OperationContext.RESERVATION:
    case OperationContext.PRODUCTION:
      return true;
    default:
      return false;
  }
}

/** @deprecated use WarehouseOperation */
export type WarehouseOperationType = WarehouseOperation;

/** Order-list bulk / quick action kinds → canonical operations. */
export type OrderListBulkActionKind =
  | "change_status"
  | "issue_document"
  | "set_priority"
  | "change_shipping"
  | "add_note"
  | "change_payment_status"
  | "custom_field_value"
  | "delete"
  | "export"
  | "print"
  | "change_tags"
  | "remove_tags";

export function orderListBulkActionToOperation(kind: OrderListBulkActionKind): WarehouseOperation {
  switch (kind) {
    case "delete":
      return "order.delete_orders";
    case "change_status":
      return "order.change_panel_status";
    case "issue_document":
      return "order.issue_document_type";
    case "set_priority":
      return "order.set_priority";
    case "change_shipping":
      return "order.change_shipping";
    case "add_note":
      return "order.add_note";
    case "change_payment_status":
      return "order.change_payment_status";
    case "custom_field_value":
      return "order.custom_field_value";
    case "export":
      return "order.export";
    case "print":
      return "order.print_documents";
    case "change_tags":
      return "order.change_tags";
    case "remove_tags":
      return "order.remove_tags";
    default:
      return "order.edit";
  }
}

export function getOrderListBulkActionPolicy(kind: OrderListBulkActionKind): OperationPolicy {
  return getOperationPolicy(orderListBulkActionToOperation(kind));
}

export function orderListBulkActionRequiresWarehouse(kind: OrderListBulkActionKind): boolean {
  return getOrderListBulkActionPolicy(kind).requiresWarehouse;
}

export const WAREHOUSE_REQUIRED_FOR_STOCK_OP_MESSAGE =
  "Wybierz magazyn realizacji, aby wykonać tę operację magazynową.";

/** @deprecated filtered_query / „wszystkie z filtra” nie wymusza magazynu dla workflow OMS */
export const WAREHOUSE_REQUIRED_FOR_FILTERED_QUERY_MESSAGE = WAREHOUSE_REQUIRED_FOR_STOCK_OP_MESSAGE;

/** Rows for audit / docs. */
export function listOperationPolicies(): Array<{ operation: WarehouseOperation } & OperationPolicy> {
  return (Object.keys(POLICY) as WarehouseOperation[]).map((operation) => ({
    operation,
    ...POLICY[operation],
  }));
}

File: backend/services/warehouse_operation_policy.py
"""
Central warehouse requirement policy — OMS workflow vs WMS/stock ops.

Keep operation ids in sync with frontend ``warehouseOperationPolicy.ts``.
"""

from __future__ import annotations

from dataclasses import dataclass
from enum import Enum
from typing import Dict, Optional


class OperationContext(str, Enum):
    NONE = "NONE"
    ORDER_WORKFLOW = "ORDER_WORKFLOW"
    INVENTORY = "INVENTORY"
    LOCATION = "LOCATION"
    DOCUMENT = "DOCUMENT"
    RESERVATION = "RESERVATION"
    PRODUCTION = "PRODUCTION"


@dataclass(frozen=True)
class OperationPolicy:
    requires_warehouse: bool
    context: OperationContext
    reason: str


_POLICY: Dict[str, OperationPolicy] = {
    # ORDER_WORKFLOW
    "order.change_panel_status": OperationPolicy(
        False, OperationContext.ORDER_WORKFLOW, "Status panelu to workflow OMS, nie operacja na stanach."
    ),
    "order.change_status": OperationPolicy(
        False, OperationContext.ORDER_WORKFLOW, "Status zamówienia to workflow OMS."
    ),
    "order.change_customer": OperationPolicy(
        False, OperationContext.ORDER_WORKFLOW, "Dane klienta nie zależą od magazynu."
    ),
    "order.change_tags": OperationPolicy(
        False, OperationContext.ORDER_WORKFLOW, "Tagi to metadane zamówienia."
    ),
    "order.remove_tags": OperationPolicy(
        False, OperationContext.ORDER_WORKFLOW, "Usuwanie tagów to metadane zamówienia."
    ),
    "order.set_priority": OperationPolicy(
        False, OperationContext.ORDER_WORKFLOW, "Priorytet to atrybut workflow."
    ),
    "order.add_note": OperationPolicy(
        False, OperationContext.ORDER_WORKFLOW, "Notatka nie modyfikuje stanów."
    ),
    "order.delete_note": OperationPolicy(
        False, OperationContext.ORDER_WORKFLOW, "Usuwanie notatki nie modyfikuje stanów."
    ),
    "order.change_payment_status": OperationPolicy(
        False, OperationContext.ORDER_WORKFLOW, "Status płatności to workflow OMS."
    ),
    "order.change_shipping": OperationPolicy(
        False, OperationContext.ORDER_WORKFLOW, "Metoda dostawy to atrybut zamówienia."
    ),
    "order.issue_document_type": OperationPolicy(
        False,
        OperationContext.ORDER_WORKFLOW,
        "Ustawienie typu dokumentu (flaga), nie wystawienie dokumentu magazynowego.",
    ),
    "order.edit": OperationPolicy(
        False, OperationContext.ORDER_WORKFLOW, "Edycja pól OMS bez ruchu towaru."
    ),
    "order.export": OperationPolicy(
        False, OperationContext.ORDER_WORKFLOW, "Eksport danych jest operacją biurową."
    ),
    "order.print_documents": OperationPolicy(
        False,
        OperationContext.ORDER_WORKFLOW,
        "Druk dokumentów handlowych nie wymaga aktywnego magazynu WMS.",
    ),
    "order.custom_field_value": OperationPolicy(
        False, OperationContext.ORDER_WORKFLOW, "Pola dodatkowe to metadane zamówienia."
    ),
    "order.delete_orders": OperationPolicy(
        False,
        OperationContext.ORDER_WORKFLOW,
        "Usuwanie/archiwizacja rekordu zamówienia to operacja administracyjna OMS.",
    ),
    "order.bulk_patch": OperationPolicy(
        False, OperationContext.ORDER_WORKFLOW, "Masowy patch pól OMS (priorytet, notatki, …)."
    ),
    "order.bulk_panel_status": OperationPolicy(
        False, OperationContext.ORDER_WORKFLOW, "Masowa zmiana statusu panelu — workflow OMS."
    ),
    # DOCUMENT
    "document.create_warehouse": OperationPolicy(
        True, OperationContext.DOCUMENT, "Dokument magazynowy jest w kontekście magazynu."
    ),
    "document.delete_warehouse": OperationPolicy(
        True, OperationContext.DOCUMENT, "Usuwanie dokumentu magazynowego wymaga magazynu."
    ),
    # RESERVATION
    "reservation.create": OperationPolicy(
        True, OperationContext.RESERVATION, "Rezerwacja blokuje stany w magazynie."
    ),
    "reservation.release": OperationPolicy(
        True, OperationContext.RESERVATION, "Zwolnienie rezerwacji dotyczy stanów."
    ),
    "reservation.delete": OperationPolicy(
        True, OperationContext.RESERVATION, "Usuwanie rezerwacji dotyczy stanów."
    ),
    # INVENTORY / WMS
    "wms.picking": OperationPolicy(True, OperationContext.INVENTORY, "Zbieranie schodzi ze stanów."),
    "wms.picking_bulk": OperationPolicy(
        True, OperationContext.INVENTORY, "Masowe zbieranie to proces WMS."
    ),
    "wms.packing": OperationPolicy(True, OperationContext.INVENTORY, "Pakowanie operuje na stanach/nośnikach."),
    "wms.receiving": OperationPolicy(True, OperationContext.INVENTORY, "Przyjęcie zwiększa stany."),
    "wms.putaway": OperationPolicy(True, OperationContext.LOCATION, "Rozlokowanie na lokalizacjach."),
    "wms.putaway_bulk": OperationPolicy(
        True, OperationContext.LOCATION, "Masowe rozlokowanie to proces WMS."
    ),
    "wms.mm": OperationPolicy(True, OperationContext.LOCATION, "MM zmienia lokalizacje i stany."),
    "wms.inventory_count": OperationPolicy(True, OperationContext.INVENTORY, "Inwentaryzacja koryguje stany."),
    "wms.relocation": OperationPolicy(True, OperationContext.LOCATION, "Rozlokowanie nośników."),
    "wms.returns_terminal": OperationPolicy(
        True, OperationContext.INVENTORY, "Terminal zwrotów przyjmuje towar na stany."
    ),
    "wms.delete_picking_batch": OperationPolicy(
        True, OperationContext.INVENTORY, "Usuwanie zbioru zbierania to artefakt WMS."
    ),
    # LOCATION
    "location.create": OperationPolicy(True, OperationContext.LOCATION, "Lokalizacja należy do magazynu."),
    "location.update": OperationPolicy(True, OperationContext.LOCATION, "Edycja lokalizacji."),
    "location.delete": OperationPolicy(True, OperationContext.LOCATION, "Usuwanie lokalizacji."),
    "location.move_stock": OperationPolicy(True, OperationContext.LOCATION, "Ruch towaru między lokalizacjami."),
    # PRODUCTION
    "production.execute": OperationPolicy(
        True, OperationContext.PRODUCTION, "Produkcja zużywa/przyjmuje stany."
    ),
    "production.reserve_materials": OperationPolicy(
        True, OperationContext.PRODUCTION, "Rezerwacja surowców na produkcję."
    ),
    "admin.warehouse_settings": OperationPolicy(
        True, OperationContext.NONE, "Konfiguracja jest per magazyn."
    ),
}


def get_operation_policy(operation: str) -> OperationPolicy:
    try:
        return _POLICY[operation]
    except KeyError as exc:
        raise KeyError(f"Unknown warehouse operation: {operation}") from exc


def requires_warehouse(operation: str) -> bool:
    return get_operation_policy(operation).requires_warehouse


def requires_warehouse_for_context(context: OperationContext) -> bool:
    return context in {
        OperationContext.INVENTORY,
        OperationContext.LOCATION,
        OperationContext.DOCUMENT,
        OperationContext.RESERVATION,
        OperationContext.PRODUCTION,
    }


def assert_warehouse_if_required(
    operation: str,
    warehouse_id: Optional[int],
    *,
    detail: Optional[str] = None,
) -> None:
    """Raise HTTPException-friendly ValueError when warehouse is required but missing."""
    policy = get_operation_policy(operation)
    if not policy.requires_warehouse:
        return
    if warehouse_id is None or int(warehouse_id) < 1:
        raise ValueError(detail or f"warehouse_id is required for {operation}: {policy.reason}")


def list_operation_policies() -> list[dict]:
    return [
        {
            "operation": op,
            "requires_warehouse": p.requires_warehouse,
            "context": p.context.value,
            "reason": p.reason,
        }
        for op, p in sorted(_POLICY.items())
    ]
